code/espurna/sensors/ADE7953Sensor.h

```diff
@@ -0,0 +1,215 @@
+            uint32_t current_rms1 = 0;
+            uint32_t current_rms2 = 0;
+            uint32_t voltage_rms = 0;            
+
+            voltage_rms = Ade7953Read(0x31C);      // Both relays
+            current_rms1 = Ade7953Read(0x31B);     // Relay 1
+            if (current_rms1 < 2000) {             // No load threshold (20mA)
+                current_rms1 = 0;
+                active_power1 = 0;
+            } else {
+                active_power1 = (int32_t)Ade7953Read(0x313) * -1;  // Relay 1
+                active_power1 = (active_power1 > 0) ? active_power1 : 0;
+            }
+            current_rms2 = Ade7953Read(0x31A);     // Relay 2
+            if (current_rms2 < 2000) {             // No load threshold (20mA)
+                current_rms2 = 0;
+                active_power2 = 0;
+            } else {
+                active_power2 = (int32_t)Ade7953Read(0x312);  // Relay 2
+                active_power2 = (active_power2 > 0) ? active_power2 : 0;
+            }
+            _voltage = (float) voltage_rms / ADE7953_UREF;            
+            
+            writeFloat(
+                ADE7953_ALL_RELAYS, 
+                (float)(current_rms1 + current_rms2) / (ADE7953_IREF * 10), 
+                (float)(active_power1 + active_power2) / (ADE7953_PREF / 10)
+            );
+            writeFloat(
+                ADE7953_RELAY_1, 
+                (float) current_rms1 / (ADE7953_IREF * 10), 
+                (float) active_power1 / (ADE7953_PREF / 10)
+            );    
+            writeFloat(
+                ADE7953_RELAY_2, 
+                (float)current_rms2 / (ADE7953_IREF * 10), 
+                (float)active_power2 / (ADE7953_PREF / 10)
+            );
+        }
+
+        void writeFloat(unsigned int relay, float current, float power) {
+            auto& reading_ref = _readings.at(relay); 
+            reading_ref.current = current;
+            reading_ref.power = power;
+        }
+
+        // Current value for slot # index
+        double value(unsigned char index) {        
+            if (index == 0) return _voltage;            
+            int relay = (index - 1) / ADE7953_TOTAL_DEVICES;	
+            index = index % ADE7953_TOTAL_DEVICES;
+            if (index == 0) return _readings[relay].energy;
+            if (index == 1) return _readings[relay].current;
+            if (index == 2) return _readings[relay].power;     
+            return 0;
+        }
+
+    protected:  
+        void _init() {                     
+            delay(100);                     // Need 100mS to init ADE7953
+            Ade7953Write(0x102, 0x0004);    // Locking the communication interface (Clear bit COMM_LOCK), Enable HPF
+            Ade7953Write(0x0FE, 0x00AD);    // Unlock register 0x120
+            Ade7953Write(0x120, 0x0030);    // Configure optimum setting       
+                    
+            _ready = true;
+        }
+
+        int Ade7953RegSize(uint16_t reg) {
```

Comment:
This might as well be static, since we don't use any member access. And we already are inside a class, Ade7953 prefix is redundant.

I was somewhat confused by the sizing function, is this mentioned somewhere in the datasheet?

*Kind-of related*
If it is really how it works, I tried making it more compact (idk about what is more readable though...)
```cpp
const int reg_size_new(const uint16_t reg) {

    const uint8_t mask = ((reg >> 8) & 0b11111111);

    if (!mask || (mask & 0b1100)) {
        return 1;
    } else if (mask & 0b0011) {
        return mask + 1;
    }

    return 0;

}
```
You can try this with ` ~/.platformio/packages/toolchain-xtensa/bin/xtensa-lx106-elf-gdb` (you'll need start building something with `espressif8266@2.2.2` platform first, it will move the old one to a separate dir)
```gdb
(gdb) disassemble reg_size
Dump of assembler code for function reg_size(unsigned short):
   0x40201018 <+0>:     extui   a2, a2, 8, 4
   0x4020101b <+3>:     movi.n  a3, 8
   0x4020101d <+5>:     bltu    a3, a2, 0x40201069 <reg_size(unsigned short)+81>
   0x40201020 <+8>:     l32r    a3, 0x40201014
   0x40201023 <+11>:    addx4   a2, a2, a3
   0x40201026 <+14>:    l32i.n  a2, a2, 0
   0x40201028 <+16>:    jx      a2
   0x4020102b <+19>:    and     a6, a3, a0
   0x4020102e <+22>:    excw
   0x40201031 <+25>:    excw
   0x40201034 <+28>:    or      a1, a0, a5
   0x40201037 <+31>:    and     a5, a5, a4
   0x4020103a <+34>:    excw
   0x4020103d <+37>:    excw
   0x40201040 <+40>:    s32i.n  a6, a0, 4
   0x40201042 <+42>:    excw
   0x40201045 <+45>:    excw
   0x40201048 <+48>:    excw
   0x4020104b <+51>:    and     a6, a3, a4
   0x4020104e <+54>:    excw
   0x40201051 <+57>:    s8i     a0, a6, 0
   0x40201054 <+60>:    andbc   b0, b12, b0
   0x40201057 <+63>:    addi.n  a2, a2, 1
   0x40201059 <+65>:    j       0x4020105e <reg_size(unsigned short)+70>
   0x4020105c <+68>:    movi.n  a2, 0
   0x4020105e <+70>:    addi.n  a2, a2, 1
   0x40201060 <+72>:    j       0x40201065 <reg_size(unsigned short)+77>
   0x40201063 <+75>:    movi.n  a2, 0
   0x40201065 <+77>:    addi.n  a2, a2, 1
   0x40201067 <+79>:    ret.n
   0x40201069 <+81>:    movi.n  a2, 0
   0x4020106b <+83>:    ret.n
(gdb) disassemble reg_size_new
Dump of assembler code for function reg_size(unsigned short):
   0x40202f24 <+0>:     extui   a3, a2, 8, 8
   0x40202f27 <+3>:     movi.n  a2, 1
   0x40202f29 <+5>:     beqz.n  a3, 0x40202f3c <reg_size_new(unsigned short)+24>
   0x40202f2b <+7>:     movi.n  a4, 12
   0x40202f2d <+9>:     and     a4, a3, a4
   0x40202f30 <+12>:    bnez.n  a4, 0x40202f3c <reg_size_new(unsigned short)+24>
   0x40202f32 <+14>:    extui   a2, a3, 0, 2
   0x40202f35 <+17>:    addi.n  a3, a3, 1
   0x40202f37 <+19>:    movnez  a4, a3, a2
   0x40202f3a <+22>:    mov.n   a2, a4
   0x40202f3c <+24>:    ret.n
```
At least what it seems like, that the code size goes down 77 bytes.

Reply:
I need to learn more about moving bytes, I'm not really sure what reg_size does, but it seems to be a very important function.

Reply:
And just to be clear about this, what happens is:
```python
>>> '{:#06b}'.format(0x31C >> 8)
'0b0011'
>>> '{:#06b}'.format(0x21C >> 8)
'0b0010'
```
Then, when checking last 2 bits
```
>>> (0b0011 & 0b0011) + 1
4
>>> (0b0010 & 0b0011) + 1
3
```
So, the 0x3... prefix contains 4 bytes => 32 bits and 0x2... contains 3 bytes => 24 bits